src/coreComponents/physicsSolvers/multiphysics/ThermalSinglePhasePoromechanicsKernel.hpp

```diff
@@ -0,0 +1,606 @@
+
+    for( integer localNode = 0; localNode < numNodesPerElem; ++localNode )
+    {
+      for( integer dim = 0; dim < numDofPerTestSupportPoint; ++dim )
+      {
+        localIndex const dof = LvArray::integerConversion< localIndex >( stack.localRowDofIndex[numDofPerTestSupportPoint*localNode + dim] - m_dofRankOffset );
+        if( dof < 0 || dof >= m_matrix.numRows() ) continue;
+
+        m_matrix.template addToRowBinarySearchUnsorted< parallelDeviceAtomic >( dof,
+                                                                                &stack.localTemperatureDofIndex,
+                                                                                stack.dLocalResidualMomentum_dTemperature[numDofPerTestSupportPoint * localNode + dim],
+                                                                                1 );
+      }
+    }
+
+    // Step 2: assemble the derivatives of mass balance residual wrt temperature into the global matrix
+
+    localIndex const massDof = LvArray::integerConversion< localIndex >( stack.localPressureDofIndex - m_dofRankOffset );
+    if( 0 <= massDof && massDof < m_matrix.numRows() )
+    {
+      m_matrix.template addToRow< serialAtomic >( massDof,
+                                                  &stack.localTemperatureDofIndex,
+                                                  stack.dLocalResidualMass_dTemperature[0],
+                                                  1 );
+    }
+
+    // Step 3: assemble the energy balance and its derivatives into the global matrix
+
+    localIndex const energyDof = LvArray::integerConversion< localIndex >( stack.localTemperatureDofIndex - m_dofRankOffset );
+    if( 0 <= energyDof && energyDof < m_matrix.numRows() )
+    {
+      m_matrix.template addToRowBinarySearchUnsorted< serialAtomic >( energyDof,
+                                                                      stack.localRowDofIndex,
+                                                                      stack.dLocalResidualEnergy_dDisplacement[0],
+                                                                      nUDof );
+      m_matrix.template addToRow< serialAtomic >( energyDof,
+                                                  &stack.localPressureDofIndex,
+                                                  stack.dLocalResidualEnergy_dPressure[0],
+                                                  1 );
+      m_matrix.template addToRow< serialAtomic >( energyDof,
+                                                  &stack.localTemperatureDofIndex,
+                                                  stack.dLocalResidualEnergy_dTemperature[0],
+                                                  1 );
+
+      RAJA::atomicAdd< serialAtomic >( &m_rhs[energyDof], stack.localResidualEnergy[0] );
+    }
+
+    return maxForce;
+  }
+
+protected:
+
+  /// The rank global density derivative wrt temperature
```

Comment:
fluid density or global density?

Reply:
Done, now the comment says "fluid density"